Special scrub handling of hinfo_key errors

dzafman opened the pull request.

dzafman commented on src/include/rados/rados_types.hpp:
```diff
@@ -55,21 +55,31 @@ struct err_t {
-    DATA_DIGEST_MISMATCH_OI = 1 << 9,
-    OMAP_DIGEST_MISMATCH_OI = 1 << 10,
-    SIZE_MISMATCH_OI        = 1 << 11,
+    DATA_DIGEST_MISMATCH_OI = 1 << 9,   // Old
```
Should I keep these old names for compilation of older external code?

dzafman commented on src/tools/rados/rados.cc:
```diff
@@ -1369,35 +1375,67 @@ static void dump_shard(const shard_info_t& shard,
+      bufferlist bl;
+      bufferlist::iterator bliter = k->second.begin();
+      decode(hi, bliter);  // Can't be corrupted
+      f.dump_stream("hashinfo") << hi;
```
Instead of using operator<< should we output JSON for the HashInfo.

jdurgin replied:
sure

dzafman commented on src/tools/rados/rados.cc:
```diff
@@ -1369,35 +1375,67 @@ static void dump_shard(const shard_info_t& shard,
+      bufferlist bl;
+      bufferlist::iterator bliter = k->second.begin();
+      decode(ss, bliter);  // Can't be corrupted
+      f.dump_stream("snapset") << ss;
```
Update this too from operator<< to JSON for SnapSet?

jdurgin replied:
yeah, sounds like a good idea

dzafman commented:
Want to also rename ss_attr_corrupted in the list-inconsistent-snapset and make sure that non-primary corruption is also reflected their.

dzafman commented:
It would be a much more involved change to handle all shards when dealing with snapshot analysis.  I filed tracker http://tracker.ceph.com/issues/23428 to at least indicate which authoritative shard was used to determine snapshot consistency.

dzafman commented:
> Make sure that non-authoritative corruption is also reflected in list-inconsistent-snapset.

This is a lot of work which I don't think is a priority right now.

jdurgin commented:
This looks good, I agree json for the snapset and hashinfo would be nice as well

tchaikov commented:
http://pulpito.ceph.com/kchai-2018-03-28_14:42:54-rados-wip-kefu-testing-2018-03-28-1948-distro-basic-smithi/

tchaikov commented:
it's good to merge if we want to improve the json representation of snapset and hashinfo in another PR.

dzafman commented:
@jdurgin I rebased this branch.  Changed the system xattrs to output json format.  Added a fix for tracker 23428.

Concerns for object_info/hashinfo/snapset:

- [ ] It is easier to visually compare the differences in the single line format.  Just use that format?
- [ ] Keep the strings using old name (e.g. "object_info") for compatibility and add "*_json" for the json?
- [ ] I noticed that the single line format doesn't show all fields
- [ ] In the json output I'd like to show the flags as an array of strings instead an integer

dzafman commented:
retest this please - run-tox-mgr-dashboard (Failed)

jdurgin commented:
making the snapset/hash_info/object_info json means it's machine-readable, instead of having to parse it again. this is worth it imo, since the single-line format isn't legible to folks who don't know the code already.

Using flag names instead of an int is a good idea. I'm fine keeping the same field names, just add something to PendingReleaseNotes about it.

So basically, this looks good to me, it just needs a release note.

dzafman commented:
@jdurgin I had to rebase because of conflicts with the PendingReleaseNotes

dzafman commented:
retest this please

tchaikov commented on PendingReleaseNotes:
```diff
@@ -126,3 +126,42 @@ method. See http://docs.ceph.com/docs/luminous/mgr/restful for details.
+    now snapset_missing).  The error snapset_mismatch has been renamed to snapset_error
+    to better reflect what it means.
+
+  * The head snapset information is output in json format as "snapset."  This means that
```
"snapset".
